pipelining_example/pipelining_example.c

```diff
@@ -0,0 +1,2793 @@
+#define _GNU_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#include <arpa/inet.h>
+#include <byteswap.h>
+#include <endian.h>
+#include <getopt.h>
+#include <infiniband/verbs.h>
+#include <infiniband/mlx5dv.h>
+#include <inttypes.h>
+#include <netdb.h>
+#include <signal.h>
+#include <time.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/socket.h>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#ifdef __x86_64__
+#define CACHE_LINE_SIZE 64
+#elif defined(__aarch64__)
+#define CACHE_LINE_SIZE 128
+#else
+#error Unknown architecture
+#endif
+#define CACHE_LINE_MASK (CACHE_LINE_SIZE - 1)
+#define CACHE_LINE_ALIGN(_x) ((_x + CACHE_LINE_MASK) & ~CACHE_LINE_MASK)
+
+/* poll CQ timeout in millisec (2 seconds) */
+#define MAX_POLL_CQ_TIMEOUT 2000
+#define MAX_WC_PER_POLL 32
+
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+static inline uint64_t htonll(uint64_t x) { return bswap_64(x); }
+static inline uint64_t ntohll(uint64_t x) { return bswap_64(x); }
+#elif __BYTE_ORDER == __BIG_ENDIAN
+static inline uint64_t htonll(uint64_t x) { return x; }
+static inline uint64_t ntohll(uint64_t x) { return x; }
+#else
+#error __BYTE_ORDER is neither __LITTLE_ENDIAN nor __BIG_ENDIAN
+#endif
+
+enum test_type {
+	TEST_TYPE_WRITE,
+	TEST_TYPE_READ,
+	TEST_TYPE_UNKNOWN,
+};
+
+struct signature_ops {
+	const char *name;
+	size_t pi_size;
+	void (*set_sig_domain)(struct mlx5dv_sig_block_domain *, void *);
+	void (*dump_pi)(void *pi);
+	uint8_t check_mask;
+};
+
+enum signature_types {
+	SIG_TYPE_CRC32 = 0,
+	SIG_TYPE_MAX,
+};
+
+void set_sig_domain_crc32(struct mlx5dv_sig_block_domain *, void *);
+void dump_pi_crc32(void *);
+
+const struct signature_ops sig_ops[SIG_TYPE_MAX] = {
+	[SIG_TYPE_CRC32] = {
+		.name		= "crc32",
+		.pi_size	= 4,
+		.set_sig_domain	= set_sig_domain_crc32,
+		.dump_pi	= dump_pi_crc32,
+		.check_mask	= MLX5DV_SIG_CHECK_CRC32,
+	},
+};
```

Comment:
The example is developed for storage protocols. And T10-DIF is a commonly used signature type used in the storage area.

Let's use T10-DIF instead of CRC32.

Since one signature type is supported, let's drop signature_ops and use T10-DIF directly.

Reply:
Ok, I will change it, thanks!